### ISSUE TITLE
ld: final link failed: bad value with power10

### ISSUE THREAD

**Issue by shenki** [linuxppc/issues#400]
I have a power10 toolchain:

GNU ld (GNU Binutils) 2.36.1
GCC 10.3.0

Building (v5.15.18 or, v5.18-rc0) for ppc64le:

```
  BOOTAS  arch/powerpc/boot/crt0.o
  BOOTAR  arch/powerpc/boot/wrapper.a
  WRAP    arch/powerpc/boot/zImage.epapr
arch/powerpc/boot/wrapper.a(crt0.o): in function `p_base':
(.text+0x150): call to `platform_init' lacks nop, can't restore toc; (toc save/adjust stub)
(.text+0x154): call to `start' lacks nop, can't restore toc; (toc save/adjust stub)
powerpc64le-buildroot-linux-gnu-ld: final link failed: bad value
```

I can also reproduce with ld 2.37. GCC 11 doesn't build the powerpc kernel due to https://github.com/linuxppc/issues/issues/388.

```
  powerpc64le-buildroot-linux-gnu-gcc -Wp,-MD,arch/powerpc/boot/.crt0.o.d -D__ASSEMBLY__ -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs -fno-strict-aliasing -O2 -msoft-float -mno-altivec -mno-vsx -pipe -fomit-frame-pointer -fno-builtin -fPIC -nostdinc -I../arch/powerpc/include -I./arch/powerpc/include/generated -I../include -I./include -I../arch/powerpc/include/uapi -I./arch/powerpc/include/generated/uapi -I../include/uapi -I./include/generated/uapi -include ../include/linux/compiler-version.h -include ../include/linux/kconfig.h -m64 -isystem /home/joel/dev/buildroot/output/host/lib/gcc/powerpc64le-buildroot-linux-gnu/10.3.0/include -mlittle-endian -mabi=elfv2 -nostdinc -c -o arch/powerpc/boot/crt0.o ../arch/powerpc/boot/crt0.S
```

Adding a nop after `bl platform_init` fixes the first error:
```
--- a/arch/powerpc/boot/crt0.S
+++ b/arch/powerpc/boot/crt0.S
@@ -219,7 +219,7 @@ p_base:     mflr    r10             /* r10 now points to runtime addr of p_base */
 #endif  /* __powerpc64__ */
        /* Call platform_init() */
        bl      platform_init
-
+       nop
        /* Call start */
        b       start
```

But doing the same after `b start` does not.

**Comment by shenki** [comment]
This patch links:

```
--- a/arch/powerpc/boot/crt0.S
+++ b/arch/powerpc/boot/crt0.S
@@ -219,9 +219,11 @@ p_base:    mflr    r10             /* r10 now points to runtime addr of p_base */
 #endif  /* __powerpc64__ */
        /* Call platform_init() */
        bl      platform_init
+       nop
 
        /* Call start */
-       b       start
+       bl      start
+       nop
 
 #ifdef __powerpc64__
 
```

**Comment by chleroy** [comment]
> 

ZZZ

**Comment by chleroy** [comment]
> This patch links:
> 
> ```
> --- a/arch/powerpc/boot/crt0.S
> +++ b/arch/powerpc/boot/crt0.S
> @@ -219,9 +219,11 @@ p_base:    mflr    r10             /* r10 now points to runtime addr of p_base */
>  #endif  /* __powerpc64__ */
>         /* Call platform_init() */
>         bl      platform_init
> +       nop
>  
>         /* Call start */
> -       b       start
> +       bl      start
> +       nop
>  
>  #ifdef __powerpc64__
>  
> ```

Maybe it should use be a branch to .start instead ?

**Comment by mpe** [comment]
It _shouldn't_ make any difference. See torvalds/linux@b1576fec7f4d.

**Comment by chleroy** [comment]
I'm surprised, I have no problem building ppc64le_defconfig with gcc 10.3 / binutils 2.36.1 from <https://mirrors.edge.kernel.org/pub/tools/crosstool/files/bin/x86_64/10.3.0/>:

```
  LD      vmlinux
  SYSMAP  System.map
  SORTTAB vmlinux
  CHKHEAD vmlinux
  CHKREL  vmlinux
  BOOTAS  arch/powerpc/boot/crt0.o
  BOOTCC  arch/powerpc/boot/cuboot.o
  BOOTCC  arch/powerpc/boot/decompress.o
  BOOTCC  arch/powerpc/boot/devtree.o
  BOOTAS  arch/powerpc/boot/div64.o
  BOOTCC  arch/powerpc/boot/elf_util.o
  BOOTCC  arch/powerpc/boot/fdt.o
  BOOTCC  arch/powerpc/boot/fdt_ro.o
  BOOTCC  arch/powerpc/boot/fdt_rw.o
  BOOTCC  arch/powerpc/boot/fdt_strerror.o
  BOOTCC  arch/powerpc/boot/fdt_sw.o
  BOOTCC  arch/powerpc/boot/fdt_wip.o
  BOOTCC  arch/powerpc/boot/inffast.o
  BOOTCC  arch/powerpc/boot/inflate.o
  BOOTCC  arch/powerpc/boot/inftrees.o
  BOOTCC  arch/powerpc/boot/libfdt-wrapper.o
  BOOTCC  arch/powerpc/boot/main.o
  BOOTCC  arch/powerpc/boot/ns16550.o
  BOOTCC  arch/powerpc/boot/ofconsole.o
  BOOTCC  arch/powerpc/boot/oflib.o
  BOOTAS  arch/powerpc/boot/opal-calls.o
  BOOTCC  arch/powerpc/boot/opal.o
  BOOTCC  arch/powerpc/boot/serial.o
  BOOTCC  arch/powerpc/boot/simple_alloc.o
  BOOTCC  arch/powerpc/boot/stdio.o
  BOOTCC  arch/powerpc/boot/stdlib.o
  BOOTAS  arch/powerpc/boot/string.o
  BOOTAS  arch/powerpc/boot/util.o
  BOOTCC  arch/powerpc/boot/epapr.o
  BOOTCC  arch/powerpc/boot/epapr-wrapper.o
  BOOTCC  arch/powerpc/boot/of.o
  BOOTAS  arch/powerpc/boot/pseries-head.o
  LDS     arch/powerpc/boot/zImage.lds
  BOOTCC  arch/powerpc/boot/empty.o
  BOOTAR  arch/powerpc/boot/wrapper.a
  WRAP    arch/powerpc/boot/zImage.pseries
  WRAP    arch/powerpc/boot/zImage.epapr
[chleroy@PO20335 linux-powerpc]$ /opt/gcc-10.3.0-nolibc/powerpc64-linux/bin/powerpc64-linux-as -v
GNU assembler version 2.36.1 (powerpc64-linux) using BFD version (GNU Binutils) 2.36.1
[chleroy@PO20335 linux-powerpc]$ /opt/gcc-10.3.0-nolibc/powerpc64-linux/bin/powerpc64-linux-gcc -v
Using built-in specs.
COLLECT_GCC=/opt/gcc-10.3.0-nolibc/powerpc64-linux/bin/powerpc64-linux-gcc
COLLECT_LTO_WRAPPER=/opt/gcc-10.3.0-nolibc/powerpc64-linux/bin/../libexec/gcc/powerpc64-linux/10.3.0/lto-wrapper
Target: powerpc64-linux
Configured with: /home/arnd/git/gcc/configure --target=powerpc64-linux --enable-targets=all --prefix=/home/arnd/cross/x86_64/gcc-10.3.0-nolibc/powerpc64-linux --enable-languages=c --without-headers --disable-bootstrap --disable-nls --disable-threads --disable-shared --disable-libmudflap --disable-libssp --disable-libgomp --disable-decimal-float --disable-libquadmath --disable-libatomic --disable-libcc1 --disable-libmpx --enable-checking=release
Thread model: single
Supported LTO compression algorithms: zlib
gcc version 10.3.0 (GCC) 

```

**Comment by shenki** [comment]
The buildroot toolchain I was using defaults to -mcpu=power10.

@mpe pointed out to me the reason the build borke is the -mcpu=power10 built kernel generates pcrel code. We don't really want that, as the the wrapper won't run on non-power10 machines.

I've got a patch that ensures the wrapper is built with -mcpu=powerpc64le, which ensures the behavior of the wrapper stays the same regardless of the toolchain's default cpu. I was attempting to test it when I discovered the issues in #402. It does resolve the issue for ppc64le though.
